ambry-api/src/main/java/com/github/ambry/quota/QuotaException.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2022 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota;
+
+
+/**
+ * Exception thrown by quota logic.
+ */
+public class QuotaException extends Exception {
+  private final boolean isRetryable;
+
+  /**
+   * @param message the exception message.
+   * @param isRetryable {@code true} if the quota operation can be tried again. {@code false} otherwise.
+   */
+  public QuotaException(String message, boolean isRetryable) {
+    super(message);
+    this.isRetryable = isRetryable;
+  }
+
+  /**
+   * Constructor for {@link QuotaManager}.
+   * @param message the exception message.
+   * @param cause {@link Throwable} object that is the actual cause of the exception.
+   * @param isRetryable {@code true} if the quota operation can be tried again. {@code false} otherwise.
+   */
+  public QuotaException(String message, Throwable cause, boolean isRetryable) {
```

Comment:
why do we have isRetryable? how do we differentiate the case when it's not retryable?

Reply:
Hold on, what is the quota exception used for? So, we have r/w bandwidth quota, which will slow the operation down, but not rejecting the operations. In this case, we would probably never throw an exception. And we have storage quota, which will reject uploads when it's 100%. And this is not retryable.

Like when we deal with GetOperation, we don't really check storage quota, so it's only about the read bandwidth quota. Therefore, we shouldn't care about if it's retryable or not, since it will only be slowed down. But when we deal with PutOperation, we now have to deal with the results from storage quota and write bandwidth quota. If write bandwidth quota is 100%, we will slow it down. If the storage quota is 100%, we will reject the request, and it's not retryable.

Shouldn't this be the case?

Reply:
Exception is never thrown for expected cases. QuotaException will be used for unexpected scenarios happening in the quota call stack (like some error through quota client, or maybe account or container name not found in the RestRequest).

The retryable is for if the exception (the unexpected scenario) can be resolved or not. For something coming from quota client maybe it can be resolved. But if we have a decision (either throttle or not), then no exception will ever be thrown. At no point we will charge for a request and also throw an exception about it. If a request is charged (or it is determined that it shouldn't be charged), then no exception will ever be thrown.